### ISSUE TITLE
Add an HTTP header to indicate and allow serviceWorker upgrading

### ISSUE THREAD

**Issue by machawk1** [oduwsdl/ipwb#157]
In #155, @ibnesayeed recommended adding an HTTP response header to the replay system to indicate the version of the ServiceWorker, e.g., `X-ServiceWorker: serviceWorker.js:v1`.

This would:
* If the server receives a request without that header then it can decide to send a page that would install the SW
* If installation fails, a message can be shown to the user that the browser is unsupported
* If possible, server might send a rewritten response as a fallback (which we have no intent of doing in IPWB though)
* If it does receive the header, but the version is old, then it can be upgraded with new logic

A versioning scheme should be used to indicate to the replay system to refresh the SW.

**Comment by ibnesayeed** [comment]
Wait, I was talking about a **REQUEST HEADER**, not a response header. And this header would be added by the SW. This is also described in the JCDL 2017 paper.

**Comment by machawk1** [comment]
@ibnesayeed Would the very large date-based versioning scheme* be sufficient for the version of the serviceWorker? Are there any disadvantages of requiring an ipwb user to refresh the serviceWorker in their browser when a new version of ipwb is deployed/accessed?

*e.g., 0.2017.5.18.1306 -> X-ServiceWorker: serviceWorker.js:v201705181306

**Comment by machawk1** [comment]
I'll have to re-read that paper. ;)

**Comment by ibnesayeed** [comment]
The length of the version number is not that significant. It is there to help identify stale versions of SW installed on the clients' machine which needs to be updated. As long as version string is **easily comparable**, I have no particular preference over incremental integers vs date based strings.

**Comment by machawk1** [comment]
The date-based integers are easily comparable and not arbitrary, so we will go with that. There are probably some other TODOs and test cases related to this. I'll add them as tickets once this is implemented.

Just to be clear, is this correct: the client sends this value to the ipwb system after the SW has been installed, replay compares vs. current SW implementation, responds to client with some mechanism to refresh the SW code if out-of-date?

**Comment by ibnesayeed** [comment]
> Just to be clear, is this correct: the client sends this value to the ipwb system after the SW has been installed, replay compares vs. current SW implementation, responds to client with some mechanism to refresh the SW code if out-of-date?

Yes, this header will be added when the SW makes the request for `fetch` call. The server needs to keep track of the latest version number of the SW (which will only be updated if the SW code changes, independent of the rest of the code changes). The same version number will be available in the SW code as a constant. If the server receives a request where that header was expected to be present, but it is not, then the server would think the request is coming from a client which does not have our SW installed. If the `User-Agent` seems like it is coming from a supported browser, then the client can be redirected to a page (preferably the home page with a note) that would install it. If the version number is smaller than the current latest then send necessary code to update the SW, preferably by redirecting to a page with a note.

**Comment by machawk1** [comment]
It would be useful to setup another endpoint for the replay system to be `/serviceWorker/` to allow extraction of this metadata. The current client-side serviceWorker otherwise has no way to identify what version of ipwb is running. With that in mind, maybe there would be use in setting up an `/self/` or `/ipwb/` endpoint for information about the system beyond the SW. This information can also be conveyed in the response headers. The version of IPWB is already displayed as an HTTP response header, e.g., `Server: InterPlanetary Wayback Replay/0.2017.06.15.0114`.

**Comment by machawk1** [comment]
XHR is deprecated with SWs, only `fetch()` can be used for AJAX, which limits the response headers we can retrieve. Because we are storing the ipwb version in a response header that fetch() cannot retrieve, attempting to do so might not be a viable way of getting the current version of the SW from within the SW JS.

**Comment by machawk1** [comment]
Something like:
```js
  function getMyVersion () {
     fetch(self.location.href)
     .then(function (resp) {
       console.log(resp.headers['Server'])
     })
  }
```
should work but per above, the response headers do not appear to be available from the `fetch()` response.

**Comment by ibnesayeed** [comment]
I don't see a reason why `fetch` won't show you the `Server` header. However, you are not using the right syntax I guess. The `fetch` API uses `Headers` construct that has getters and setters instead of dictionary like access. Try using [resp.headers.get('Server')](https://developer.mozilla.org/en-US/docs/Web/API/Headers/get) instead.

**Comment by machawk1** [comment]
@ibnesayeed The above was quasi-code for reference when I moved between machines (hence "like"). I am able to get the value using `resp.headers.get('Server')` as you suggested but have yet to integrate it in. Thanks!
